### DISCUSSION TITLE
Unable to Recognize Already Opened LUKS Container in Anaconda Installer

### DISCUSSION THREAD

**Discussion by AGolz** [rhinstaller/anaconda#5552] [General]
I am writing to report an issue I encountered while trying to install RockyLinux (Rocky-9.3-x86_64) using the Anaconda installer. 

**Problem Description:**
During the installation process, Anaconda is unable to recognize an already opened LUKS container.

In the text mode of the Rocky 9.3 installation, when trying to assign mount points for partitions manually to an encrypted LUKS partition, Anaconda throws the error "The existing unlocked LUKS device cannot be used for the installation without an encryption key specified for this device. Please restart the storage", I tried to work around the error with this [instruction](https://linuxconfig.org/how-to-install-fedora-rhel-centos-via-kickstart-on-an-existing-luks-device), but I ran into the following problem: the ks.cfg file successfully opens the LUKS partition, after partitioning the disk, the error "unlocked LUKS device cannot be used" does not occur, but when trying to continue the installation, an error appears on the screen: "luks device not configured" at the same time lsblk commands reflect the LUKS section as open. 

Installation File: https://raw.githubusercontent.com/AGolz/Born2beRoot/main/updates.img

**Resolution Attempted:**
I found discussions related to this issue on GitHub:
https://github.com/rhinstaller/anaconda/pull/2424
https://bugzilla.redhat.com/show_bug.cgi?id=1772902

Unfortunately, none of the solutions provided in these discussions worked for Rocky-9.3-x86_64. It's worth noting that there was no such problem in Rocky-9.1-x86_64. Detailed explanations and recommendations for version 9.1 can be found here https://github.com/AGolz/Born2beRoot, and they have been effective for version 9.1 thus far.

**Expected Behavior:**
Anaconda should recognize and utilize the already open LUKS container for installation.

I kindly request your assistance in resolving this issue. If there are any workarounds or steps I can take to successfully complete the installation, please let me know.

**Comment by h908714124** [comment]
I can confirm this problem. I think the kickstart `mount` command was introduced to help in the "keep home partition" scenario. But apparently it fails when there are LUKS partitions. Dumping what I found so far.

Here's a basic kickstart. When running it _again_, `mount` works as intended, preserving `/home`:
https://gist.github.com/h908714124/6e3b46b72e3aadf8379e1114f29e320e

Here's a kickstart that uses encryption. Installing it works _once_, but fails when we install _again_.
https://gist.github.com/h908714124/a0d26d5e6fd84f78abf3de86d17e720c

lsblk output when the `%pre`-script ends (_before_ the anaconda error):

```
NAME                                          MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
loop0                                           7:0    0 705.5M  1 loop  /run/rootfsbase
sda                                             8:0    0    20G  0 disk  
|-sda1                                          8:1    0     1G  0 part  
|-sda2                                          8:2    0   200M  0 part  
`-sda3                                          8:3    0  18.8G  0 part  
  `-luks-f2f2d49c-d543-4706-86c2-707f4741a8ce 252:0    0  18.8G  0 crypt 
    |-vgroot-home                             252:1    0     1G  0 lvm   
    `-vgroot-root                             252:2    0     4G  0 lvm   
sr0                                            11:0    1 911.7M  0 rom   /run/install/repo
zram0                                         251:0    0   3.8G  0 disk  [SWAP]
```

These kickstart commands are generated when running luks.cfg _again_:

```
mount /dev/sda1 /boot/efi
mount --reformat=ext4 /dev/sda2 /boot
mount --reformat=ext4 /dev/mapper/vgroot-root /
mount /dev/mapper/vgroot-home /home
```

Anaconda error message:

```
Running pre-installation scripts
Starting automated install...
Saving storage configuration...
Checking storage configuration...
The existing unlocked LUKS device sda3 cannot be used without an encryption key specified for this device. Please, rescan the storage.
```

Note that `sda3` is mentioned in the error message. Yet none of the `mount` commands above relate to `sda3`. Anaconda should not care about `sda3` at all. The documentation for the `mount` command says:

> The installer doesn’t necessarily have to know all the details about of the given device.

The fact that `/dev/mapper/vgroot-home` and `/dev/mapper/vgroot-root` happen to live inside `sda3` _should_ be considered such a "detail". The installer _should_ not care about this detail.

lsblk output _after_ the anaconda error:

```
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0    7:0    0 705.5M  1 loop /run/rootfsbase
sda      8:0    0    20G  0 disk 
├─sda1   8:1    0     1G  0 part 
├─sda2   8:2    0   200M  0 part 
└─sda3   8:3    0  18.8G  0 part 
sr0     11:0    1 911.7M  0 rom  /run/install/repo
zram0  251:0    0   3.8G  0 disk [SWAP]
```

Kickstart documentation:
https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html

**Comment by h908714124** [comment]
After applying the workaround (removing the `verify_unlocked_devices_have_key` check), you get a different error "luks device not configured". This one comes from blivet.

Continuing here: https://github.com/h908714124/anaconda/issues/1

**Comment by h908714124** [comment]
Update: There is [pflaster](https://github.com/jbock-java/pflaster) now, which is an elaborate kickstart pre-script that skips blivet entirely. By selecting the included "luks_lvm" storage module, or writing your own, it should always be possible work around the "existing unlocked LUKS device cannot be used for the installation without an encryption key" error.
